mlx/ops.cpp

```diff
@@ -2222,6 +2222,176 @@ array matmul(
+  // n_kv_heads > 1, n_heads % n_kv_heads == 0. K, V tensors must have same
+  // number of heads.
+
+  // Q, K must have matching last dims (d_k);
+  if (q.shape(-1) != k.shape(-1)) {
+    std::ostringstream msg;
+    msg << "[fast_inference_sdpa] query, keys expected to have matching d_k; found query "
+        << q.shape() << " for keys " << k.shape() << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // head_dim must be 128.
+  constexpr const int supported_head_dim = 128;
+  if (q.shape(-1) != supported_head_dim) {
+    std::ostringstream msg;
+    msg << "[fast_inference_sdpa] required: args.head_dim == 128 ; found "
+        << q.shape(-1) << " .";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // K, V must have matching head dims (n_kv_heads);
+  size_t n_q_heads = q.shape(-3);
+  size_t n_kv_heads = k.shape(-3);
+
+  if (k.shape(-3) != v.shape(-3)) {
+    std::ostringstream msg;
+    msg << "[fast_inference_sdpa] keys, values expected to have matching n_kv_heads; found keys with n_heads "
+        << k.shape(-3) << " for values with n_heads " << v.shape(-3) << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // n_heads % n_kv_heads == 0; n_heads >= 1, n_kv_heads >= 1.
+  if (n_q_heads % n_kv_heads != 0) {
+    std::ostringstream msg;
+    msg << "[fast_inference_sdpa] n_heads must be a multiple of n_kv_heads, found n_heads "
+        << n_q_heads << " for n_kv_heads " << n_kv_heads << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // q_seq_len == 1 for now in decoding (no mask) implementation
+  if (q.shape(-2) > 1 && !m.has_value()) {
+    std::ostringstream msg;
+    msg << "[fast_inference_sdpa] query sequence length must be 1 at decode time, found  "
+        << q.shape(-2) << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // Type promotion: common type between q, k, v;
+  auto out_type = promote_types(q.dtype(), k.dtype());
+  auto out_type_kv = promote_types(k.dtype(), v.dtype());
+  auto final_type = promote_types(out_type, out_type_kv);
+
+  if (q.dtype() != out_type) {
+    q = astype(q, out_type, s);
+  }
+  if (k.dtype() != out_type) {
+    k = astype(k, out_type, s);
+  }
+
+  if (v.dtype() != final_type) {
+    v = astype(v, final_type, s);
+  }
+
+  auto out_shape =
+      std::vector<int>({q.shape(0), q.shape(1), q.shape(2), v.shape(-1)});
+
+  auto prompt_encoding_with_or_without_tiling = [](const array& q_tensor,
+                                                   const array& k_tensor,
+                                                   const array& v_tensor,
+                                                   const array& mask_tensor,
+                                                   const float alpha,
+                                                   const Dtype out_dtype,
+                                                   StreamOrDevice s) {
+    auto q_scaled = multiply(array(alpha), q_tensor, s);
```

Comment:
Use `array(alpha, q_tensor.dtype())` to avoid an implicit upcast if `q_tensor` is a half type.

Reply:
nice, thanks!

Reply:
done